[기본 레이아웃 컴포넌트] 요술토끼(김의천) 미션 제출합니다.

wzrabbit opened the pull request:
# 🎯 1단계. 기본 레이아웃 컴포넌트 체크리스트

## 공통
- [x] 의미있는 커밋 메시지 작성
- [x] PR에 관련 없는 변경 사항이 포함 유무 확인

## Container 컴포넌트
- [x] `minWidth`, `maxWidth` 속성 적용
- [x] npm 배포
- [x] README.md 작성
- [x] 테스트 코드를 작성하였나요? (선택 사항)

## Grid 컴포넌트
- [x] `rows`, `columns`, `gap` 속성 적용
- [x] npm 배포
- [x] README.md 작성
- [x] 테스트 코드를 작성하였나요? (선택 사항)

## Flex 컴포넌트
- [x] `direction`, `justify`, `align`, `gap` 속성 적용
- [x] npm 배포
- [x] README.md 작성
- [x] 테스트 코드를 작성하였나요? (선택 사항)

## 📌 구현한 내용 설명

## 🌟 npm 배포 주소
- [주소](https://www.npmjs.com/package/amethyst-layout)
- 이름은 `amethyst-layout` 으로 지어 보았어요! 저번에 모달 미션을 할 때는 모달이 마치 백설기를 닮아서 **baeksulgi**로 아무 생각 없이 이름을 지었는데, 이번에도 아무 생각 없이 이름을 지었네요. 나중에 라이브러리를 실제로 출시해 본다면 보석 테마로 만들면 멋질 것 같아서 그냥 이름을 이렇게 지어봤어요

## 📚 스토리북 배포 주소

- [주소](https://wzrabbit.github.io/layout-component/)

## 📜  라이브러리 문서

- [주소](https://github.com/wzrabbit/layout-component/tree/amethyst#readme)

# 📌 구현한 내용 설명

안녕하세요, 도담! 요술토끼입니다.
급하게 작업하느라 스토리북 배포도 엉망이고 정신이 없네요, 좀 더 일찍 시작해 볼 걸 그랬네요. 아래는 컴포넌트를 구현할 때 중점적으로 생각했던 점입니다!


- 컴포넌트를 구현할 때는 **사용자가 자주 쓸만한 속성들** 위주로 별도의 `prop` 을 제공해 해당 속성들에 대해 빠르게 설정하여   컴포넌트를 쉽게 사용할 수 있도록 하는 것을 주 목표로 삼았습니다.
- 자주 사용되지 않는 속성이라도 사용자가 경우에 따라 추가로 사용하고 싶은 속성이 있을 것입니다. 예를 들어, `Grid` 컴포넌트에 `border-radius` 를 적용하는 것이요. 이러한 상황에도 유연하게 대응할 수 있도록 모든 `type` 에 `HTMLDivElement` 타입을 포함하여, 필요 시 사용자가 특정 속성을 추가할 수 있도록 하였습니다.
- 컴포넌트를 테스트할 겸 사용자를 위한 라이브러리 문서 작성을 위해 Storybook을 작성하였습니다. 결과물을 직접 사용자가 조작해 볼 수 있는 점이 큰 장점이라 생각합니다.
- 접근성의 경우 `Grid` 컴포넌트에는 `role="grid"` 를 적용하고, `Flex` 컴포넌트에는 `role="list"` 를 적용하였습니다만, 이외에는 어떻게 접근성을 개선할 지 모르겠어서 이 정도로만 작업했습니다. `tabIndex` 등을 작업하여 스크린리더 이용자가 `tab` 키를 누르는 것으로 요소들을 자유롭게 이동할 수 있게 만들면 좋겠지만, 자식 요소들을 직접 건드릴 방법은 떠오르지 않았습니다.

미숙한 작업물이지만 잘 부탁드립니다! 하시고 싶은 조언이 있으시다면 부디 자유롭게 해 주세요. 도담의 멋진 리뷰를 기다리겠습니다!  


<img src="https://user-images.githubusercontent.com/87642422/233928933-505b06b9-5460-4d73-bf1b-4addd84c9e9a.png" width="70px" />

wzrabbit committed:
📦 chore: 라이브러리 배포를 위한 개발환경 세팅 및 테스트 컴포넌트 구현

wzrabbit committed:
📦 chore: 스토리북 설치 및 테스트 컴포넌트 작성

wzrabbit committed:
📦 chore: styled-components 제거

- styled components를 의존성 라이브러리에 추가하는 것이 싫었음
- 설치하지 않더라도 module css 등의 방법으로 충분히 구현할 수 있을 것으로 예상

wzrabbit committed:
📦 chore: module css 사용을 위한 타입 정의 및 설정 진행

wzrabbit committed:
✨ feat: Container 컴포넌트 구현

wzrabbit committed:
✨ feat: Container 컴포넌트의 padding을 verticalPadding과 horizontalPadding으로 세분화

wzrabbit committed:
✨ feat: Grid 컴포넌트 구현

wzrabbit committed:
✨ feat: Flex 컴포넌트 구현

wzrabbit committed:
📦 chore: storybook을 github-pages로 배포할 수 있는 스크립트 작성

wzrabbit committed:
✨ feat: Grid, Flex 컴포넌트의 width, height를 설정할 수 있도록 개선

wzrabbit committed:
📦 chore: storybook 배포 스크립트의 디렉토리 변경

wzrabbit committed:
🛠️ fix: 잘못된 배포 스크립트 수정

wzrabbit committed:
📚 docs: 라이브러리 설명을 위한 README 파일 작성

wzrabbit committed:
📦 chore: 버전 업

D0Dam commented on src/lib/components/Container/Container.stories.tsx:
```diff
@@ -0,0 +1,60 @@
+  component: Container,
+  argTypes: {
+    children: {
+      description: 'The children to display',
```
오! description 이라는 건 이렇게 적용하는거였군요~! 조씁니다!👍

D0Dam commented on src/lib/components/Container/Container.tsx:
```diff
@@ -0,0 +1,44 @@
+import { HTMLAttributes, ReactNode } from 'react';
+import styles from './Container.module.css';
+
+type ContainerProps = HTMLAttributes<HTMLDivElement> & {
```
오호! 요술토끼는 type을 쓰시네요~! 저는 interface를 애용하거든요! 아무튼 험험..

요술토끼는 기본적으로 HTMLDivElement의 속성을 받도록 하고 있네요! 여기에서 질문 드립니다~!

1. 만약에 Container를 div가 아닌 다른 태그를 사용한다면 어떤 타입을 사용할 수 있을까요?
2. 기본적으로 div 태그를 Container의 기본 태그로 하고 사용자가 다른 태그로 변경할 수 있게 할 수 있지 않을까요?

꼭 적용은 안해주셔도 됩니다! 저는 저런 고민을 해보았거든요~! 제 생각을 공유하고 갑니다~!

wzrabbit replied:
와 이건 어려워서 삽질해봤는데 시간 내에 할 수 없어 보이네요.
프론트엔드는 기능 구현하는 것보다 타입 관리하는 게 세 배는 더 어려운 것 같습니다.

사용자 입장에서 `div` 대신 `header`, `section` 등과 같은 유의미한 시멘틱 태그를 사용하게 할 수 있다면 더 좋아보입니다.  따라서 구현만 할 수 있다면 이렇게 개선해 보자는 도담의 의견에는 동의합니다.

생각해 보니까 이거 2단계까지 넘어가는군요, 화요일까지 고민해보고 해결책을 찾았다면 반영해 보고, 아니면 그대로 `div` 만이라도 사용할 수 있도록 해 제출해 볼 것 같군요.

D0Dam commented on src/lib/components/Container/Container.tsx:
```diff
@@ -0,0 +1,44 @@
+    horizontalPadding,
+    children,
+    ...rest
+  } = props;
```
`verticalPadding, horizontalPadding` !! 좋은데요? 실제로 쓸 것 같은 기능을 넣어주셨군요~!

지금 생각해 보았는데 이런것도 있으면 좋을 것 같아요~!
기본적으로 Container의 height는 안의 요소에 딱 맞게 설정이 될 것 같아요.
어떤 경우에는 Container의 height가 창의 높이 전체를 채우게 할 수 있으면 좋을 것 같지 않나요?

D0Dam commented on src/lib/components/Container/Container.tsx:
```diff
@@ -0,0 +1,44 @@
+import styles from './Container.module.css';
+
+type ContainerProps = HTMLAttributes<HTMLDivElement> & {
+  maxWidth: number;
```
maxWidth는 꼭 지정해 주어야 하는 걸까요? 지정하지 않는다면 100%를 default로 넣어주는 건 어떤 것 같나요?

wzrabbit replied:
도담의 코멘트를 보고 지금은 넣을 생각이 생겨서 넣을 예정입니다. 다만 구현 당시에는 일부러 `maxWidth` 만을 필수값으로 지정하였는데, 이는 `Container` 를 이용하는 개발자 입장에서는 내용의 **가운데 정렬을 위해** 내용을 넣고 가운데 정렬이 동작하도록 그 내용의 크기를 직접 결정해 주는 정도는 해야 한다고 느꼈습니다. `maxWidth` 를 optional로 두면 값이 없을 때 `100%` 의 크기를 차지하게 될텐데, 이 경우 가운데로 정렬하는 기능을 사실상 수행하지 않은 것과 다름없다고 생각하였기 때문입니다.

요약하면 `Container` 가 동작하기 위해서는 `maxWidth` 를 직접 받는 것이 필수라 생각하여 이 값은 필수로 두었었습니다.
지금은 어떻게 사용할 지는 개발자에게 맡겨도 상관 없을 거라 생각하여 이 값은 optional로 바꾸었습니다. 사용하는 측에서 동적인 컴포넌트를 구현하기 위해 필요에 따라 `maxWidth` 를 일부러 안 넣는 경우도 있을 것 같아보이고요!

D0Dam commented on src/lib/components/Container/Container.tsx:
```diff
@@ -0,0 +1,44 @@
+  } = props;
+
+  const containerStyles = {
+    paddingTop: `${verticalPadding}px`,
+    paddingBottom: `${verticalPadding}px`,
+    paddingLeft: `${horizontalPadding}px`,
+    paddingRight: `${horizontalPadding}px`,
```
위에서 보니 이건 선택 값이던데, 값을 넣지 않으면 그냥 'px' 가 들어가는 걸까요? 

뭔가 사용할 때는 영향은 없겠지만 저는 이렇게 써 주고 싶은 욕구가 드는군요..ㅎㅎ

```suggestion
    paddingTop: `${verticalPadding || 0}px`,
    paddingBottom: `${verticalPadding || 0}px`,
    paddingLeft: `${horizontalPadding || 0}px`,
    paddingRight: `${horizontalPadding || 0}px`,
```

이거 뭔가 1줄로도 줄일 수 있을 것 같은데.. 힘드려나요? 이런식으로요!

```suggestion
    padding: `${verticalPadding || 0}px` `${horizontalPadding || 0}px`,
```

wzrabbit replied:
오 좋은데요? 당장 이렇게 고쳐야겠어요!
감사합니다!!

D0Dam commented on src/lib/components/Flex/Flex.tsx:
```diff
@@ -0,0 +1,74 @@
+
+  const flexSizes: CSSProperties = {
+    width: elementSizes.getWidth(width),
+    height: elementSizes.getHeight(height),
```
오오~! 이건 어떤 함수죠? 신기하네요~!

D0Dam commented on src/lib/components/Flex/Flex.tsx:
```diff
@@ -0,0 +1,74 @@
+  wrap?: FlexWrap;
+  gap?: number;
+  rowGap?: number;
+  columnGap?: number;
```
gap과 rowGap, columnGap을 동시에 줄 수 있는 건 무언가..! 위험해 보이네요! 대부분 그럴 일 없겠지만, gap과 columGap을 둘 다 쓰면 어떤 gap이 적용될 지 라이브러리 사용자들은 쉽게 알 수 없을 거에요!

무언가 gap을 쓰면 rowGap과 columGap을 못쓰게, rowGap과 columGap을 쓰면 gap을 못쓰게 할 수는 없을까요?
저라면 gap을 다음과 같은 타입 정의로 사용해 로직을 설계할 것 같아요!

```
interface Gap {
  row: number;
  column: number;
}

gap: number | Gap;
```

어떤 것 같나요?

wzrabbit replied:
오 이런 방법도 있어 보이는군요
저는 따로 `Gap` 타입을 선언해 따로 분리해서 작업해 보았습니다

그 외에 `CSSProperties` 라는 녀석을 이용하면 타입들을 일일이 노가다로 적어야 하는 불편함을 해결할 수 있을 것으로 보여 적용해 보았습니다!

D0Dam commented on src/lib/components/Flex/Flex.tsx:
```diff
@@ -0,0 +1,74 @@
+  verticalPadding?: number;
+  horizontalPadding?: number;
+  width?: number | string;
+  height?: number | string;
```
string은 ~~~px, ~~~%로 받기 위함인가요??
이런 식으로 써볼 수도 있답니다~!

```
type Pixel = `${number}px`;
type Persent = `${number}%`;
// vh나 vw도 이런 느낌으로

width: number | Pixel | Persent;
```
이런 느낌..?

D0Dam commented on src/lib/components/Flex/Flex.tsx:
```diff
@@ -0,0 +1,74 @@
+  };
+
+  const flexStyles: CSSProperties = {
+    gap: `${gap ?? 0}px`,
```
요술토끼는 당연히 알 것 같은 깜짝 문제입니다!

`gap ?? 0` 과 `gap || 0` 는 어떤 차이일까요?!

wzrabbit replied:
`||` 는 앞에 있는 값이 truthy하기만 하다면 그 값을 반환합니다.
`??` 는 앞에 있는 값이 `undefined` 이거나 `null` 인 상황이 아니라면 그 값을 반환합니다.

위 상황에서는 어느 쪽을 사용해도 동일하게 동작할 것으로 예상됩니다만 `??` 가 해당 값이 "정의되지 않은 - `undefined | null`" 값인 경우 뒤의 값을 선택하겠다는 의도가 더 명확하게 드러나는 것 같아 `??` 를 사용했습니다.

D0Dam commented on src/lib/components/Flex/Flex.tsx:
```diff
@@ -0,0 +1,74 @@
+    <div
+      className={styles.flex}
+      style={{ ...flexSizes, ...flexStyles }}
+      role="list"
```
👍

D0Dam commented on src/lib/components/utils/elementSizes.ts:
```diff
@@ -0,0 +1,25 @@
+
+    return 'auto';
+  },
+};
```
오호 이런 함수였군요~! : )

D0Dam commented on src/lib/components/Grid/Grid.tsx:
위의 리뷰들에서 말한 내용과 겹치는 것 같아요..!!

D0Dam reviewed (changes requested):
>이름은 amethyst-layout 으로 지어 보았어요! 저번에 모달 미션을 할 때는 모달이 마치 백설기를 닮아서 baeksulgi로 아무 생각 없이 이름을 지었는데, 이번에도 아무 생각 없이 이름을 지었네요. 나중에 라이브러리를 실제로 출시해 본다면 보석 테마로 만들면 멋질 것 같아서 그냥 이름을 이렇게 지어봤어요.

오오~~!! 이름 마음에 드는데요? 저는 왜 이런 생각을 못해봤죠? 요술토끼의 상상력에 감탄하고 갑니다 ㅎㅎ

> 자주 사용되지 않는 속성이라도 사용자가 경우에 따라 추가로 사용하고 싶은 속성이 있을 것입니다. 예를 들어, Grid 컴포넌트에 border-radius 를 적용하는 것이요. 이러한 상황에도 유연하게 대응할 수 있도록 모든 type 에 HTMLDivElement 타입을 포함하여, 필요 시 사용자가 특정 속성을 추가할 수 있도록 하였습니다.

오호~! HTMLDivElement을 포함하면 필요 시에 사용자가 원하는 속성을 추가 할 수 있나요? HTMLElement를 포함하는 것과 어떤 차이가 있을까요?
이와 관련해서는 밑에 리뷰에서도 한번 언급한 게 있으니 많은 관심 부탁드려요~! 😆

> 컴포넌트를 테스트할 겸 사용자를 위한 라이브러리 문서 작성을 위해 Storybook을 작성하였습니다. 결과물을 직접 사용자가 조작해 볼 수 있는 점이 큰 장점이라 생각합니다.

스토리북! 아주 flex와 grid 컴포넌트가 이쁘던데요? 멋집니다~!!!

> 접근성의 경우 Grid 컴포넌트에는 role="grid" 를 적용하고, Flex 컴포넌트에는 role="list" 를 적용하였습니다만, 이외에는 어떻게 접근성을 개선할 지 모르겠어서 이 정도로만 작업했습니다.

오오~! 전 이러한 부분 조차도 챙겨주지 못했는데, 요술토끼 덕분에 이마를 탁 쳤습니다! 이런 부분 챙겨주는거 너무나도 좋은 것 같아요~! 

>  tabIndex 등을 작업하여 스크린리더 이용자가 tab 키를 누르는 것으로 요소들을 자유롭게 이동할 수 있게 만들면 좋겠지만, 자식 요소들을 직접 건드릴 방법은 떠오르지 않았습니다.

기본적으로 요술토끼가 HTMLDivElement의 속성을 사용할 수 있게 컴포넌트를 설계해 놓아서 부모에 대한 접근성은 외부에서 사용자가 원한다면 잘 적용할 수 있을 것 같아요!
그리고 자식 요소에 대한 접근성도 굳이..! 상위 부모 컴포넌트에서 조절하지 않아도 된다고 생각해요! 그 자식 자체에서도 충분히 줄 수 있으니까요! (물론 이게 보기에 불편할 수도 있겠지만 말이죠..ㅎㅎ)
접근성 부분에 대해서 요술토끼가 깊게 고민한 것 같아서 저도 한 번 상기 시키고 갑니다! 좋은 고민 같이 나누어서 참 좋았어요!

우선 리뷰는 제 의견과 이렇게도 해볼 수 있지 않을까..? 와 같은 식으로 드려보았어요! 
꼭 반영하지 않아도 된답니다! 요술토끼도 제 생각에 대해 어떻게 생각하는지 한 번 생각을 나누어 보고 고치면 좋겠다고 생각한 부분에 대해서는 조금 수정해보는 시간을 가져보아요~!!

스토리북, npm 배포, readme.. 전부 깔끔 좋았습니다! (영어가 읽기 힘들었지만요..ㅎㅎ)
리뷰 확인해주시고 re-request 보내주세요~! 수고하셨습니다~~~!!!

wzrabbit committed:
♻️ refactor: padding 스타일 단순화

wzrabbit committed:
♻️ refactor: 타입 고도화

- row-gap과 column-gap, 그리고 gap을 동시에 사용할 수 없도록 변경
- CSSProperties를 사용하여 타입 정의
- Size 타입의 고도화

wzrabbit committed:
♻️ refactor: 불필요한 테스트 컴포넌트 제거

wzrabbit committed:
📦 chore: 버전 업 (v0.4.0 -> v.0.4.1)

wzrabbit reviewed (commented):
안녕하세요 도담~ 리뷰해 주셔서 감사합니다!
사용자가 불편해 할만한 점이나 컴포넌트의 엉성한 부분 등을 잘 꼬집어 주셨네요

가능한 한 많이 코멘트를 반영하고, 제 의견이 있는 부분은 의견을 적어보았습니다

사용자가 원하는 태그를 컴포넌트에 명시하면 그에 맞게 타입을 조정해 주는 부분은 저도 어려워서 접근하지 못했는데, 혹시 알고 계신 팁이 있다면 알려주시면 감사하겠습니다
타입 다루는 게 구현하는 것보다 세 배는 어려운 것 같군요 ㅋㅋ

감사합니다!

D0Dam commented:
요술토끼! 잘 해주셨네요~! 제가 예비군에 와있어서 우선은.. approve 머지를 해 놓을게요! 2단계에서 좀 더 상세하게 내용을 추가적으로 다뤄보아요! 수고하셨어요~!
